shared/middleware/helpers/sql_table.rb

```diff
@@ -0,0 +1,180 @@
+      updated_at: DateTime.now
+    })
+  end
+
+  def ensure_metadata
+    create_metadata unless metadata
+    metadata_dataset
+  end
+
+  def delete(id)
+    delete_count = items.where(row_id: id).delete
+    raise NotFound, "row `#{id}` not found in `#{@table_name}` table" unless delete_count > 0
+    true
+  end
+
+  def delete_all()
+    items.delete
+    @metadata_dataset.delete if metadata_dataset
+  end
+
+  def rename_column(old_name, new_name, ip_address)
+    ensure_metadata()
+    column_list = JSON.parse(metadata[:column_list])
+    new_column_list = TableMetadata.rename_column(column_list, old_name, new_name)
+    metadata_dataset.update({column_list: new_column_list.to_json})
+
+    items.each do |r|
+      # We want to preserve the order of the columns so creating
+      # a new hash is required.
+      new_value = {}
+      value = JSON.load(r[:value])
+      value.each do |k, v|
+        if k == old_name
+          new_value[new_name] = v
+        else
+          new_value[k] = v
+        end
+      end
+      update(r[:row_id], new_value, ip_address)
+    end
+  end
+
+  def add_column(column_name)
+    ensure_metadata()
+    column_list = JSON.parse(metadata[:column_list])
+    new_column_list = TableMetadata.add_column(column_list, column_name)
+    metadata_dataset.update({column_list: new_column_list.to_json})
+  end
+
+  def delete_column(column_name, ip_address)
+    ensure_metadata()
+    column_list = JSON.parse(metadata[:column_list])
+    new_column_list = TableMetadata.remove_column(column_list, column_name)
+    metadata_dataset.update({column_list: new_column_list.to_json})
+
+    items.each do |r|
+      value = JSON.load(r[:value])
+      value.delete(column_name)
+      update(r[:row_id], value, ip_address)
+    end
+  end
+
+  def fetch(id)
+    row = items.where(row_id: id).first
+    raise NotFound, "row `#{id}` not found in `#{@table_name}` table" unless row
+    JSON.load(row[:value]).merge('id' => row[:row_id])
+  end
+
+  def insert(value, ip_address)
+    raise ArgumentError, 'Value is not a hash' unless value.is_a? Hash
+
+    row = {
+      app_id: @channel_id,
+      storage_id: @storage_id,
+      table_name: @table_name,
+      value: value.to_json,
+      updated_at: DateTime.now,
+      updated_ip: ip_address,
+    }
+
+    tries = 0
+    begin
+      row[:row_id] = next_id
+      row[:id] = @table.insert(row)
+    rescue Sequel::UniqueConstraintViolation
+      retry if (tries += 1) < 5
+      raise
+    end
+    ensure_column_metadata(value)
+
+    JSON.load(row[:value]).merge('id' => row[:row_id])
+  end
+
+  def ensure_column_metadata(row)
+    ensure_metadata
+    column_list = JSON.parse(metadata[:column_list])
+    row_columns = TableMetadata.generate_column_list([row])
+    new_column_list = column_list.dup.concat(row_columns).uniq
+
+    if column_list != new_column_list
+      @metadata_dataset.update({column_list: new_column_list.to_json})
+    end
+  end
```

Comment:
Could you add a brief comment to this method? e.g. "Ensure that this row's columns are present in our copy of the column metadata"